### PR TITLE
Fix formatting on JsxElement/JsxExpression

### LINKED ISSUE #6716: TSX: Space is added before closing parenthesis
When I format document a space is added before closing parenthesis.

image

It should not be added.

### LINKED ISSUE #8817: JSX formatting inside of tag
**TypeScript Version:** 

1.8.9

**Code**

```ts
// A self-contained demonstration of the problem follows...
<div>{integer}<span>,{fraction}</span></div>
```

**Expected behavior:**
Same as in example

**Actual behavior:** 
```ts
<div>{integer}<span>, {fraction}</span></div>
```

Formatter inserts space after the comma inside tags

### PR DESCRIPTION
**Pull request by SaschaNaz**
Fixes #8817, fixes #6716

Input:

```tsx
<div>,{integer}</div>;
<div>,   {integer}</div>;
<span>)</span>;
<span>)   </span>;
<Router routes={                3    } />;
<Router routes={                (3)    } />;
```

Current:

```tsx
<div>, {integer}</div>;
<div>, {integer}</div>;
<span>) </span>;
<span>) </span>;
<Router routes={                3    } />;
<Router routes={                (3) } />;
```

Fix:

```tsx
<div>,{integer}</div>;
<div>,   {integer}</div>;
<span>)</span>;
<span>)   </span>;
<Router routes={3} />;
<Router routes={(3)} />;
```

With new `InsertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces` option: 

```tsx
<div>,{ integer }</div>;
<div>,   { integer }</div>;
<span>)</span>;
<span>)   </span>;
<Router routes={ 3 } />;
<Router routes={ (3) } />;
```

### CODE REVIEW

**Comment by jwbay** [comment]
Would this change
`<Router routes={ 3 } />;`
to
`<Router routes={3} />;` ?

My team has adopted the first via a lint rule because the spacing makes it easier to read (subjective opinion, I know). If the TS formatter starts enforcing the second, well, there goes that 😅

**Comment by SaschaNaz** [comment]
@jwbay So we need an additional formatting option. Thanks for the information :+1:

**Comment by SaschaNaz** [comment]
Now we have the option!

**Comment by RyanCavanaugh** [comment]
:+1: greatly appreciated! :trophy:

File: src/services/services.ts [hunk 1]
```diff
@@ -1267,6 +1267,7 @@ namespace ts {
         InsertSpaceAfterOpeningAndBeforeClosingNonemptyParenthesis: boolean;
         InsertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets: boolean;
         InsertSpaceAfterOpeningAndBeforeClosingTemplateStringBraces: boolean;
+        InsertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces: boolean;
```
Comment (mhegazy) [inline src/services/services.ts]: this needs to be optional, or it is a breaking change.

**Comment by mhegazy** [comment]
Thanks @SaschaNaz. one comment about the optionality of the new format option, and we are good to go.

**Comment by chicoxyzzy** [comment]
Does it also fix https://github.com/Microsoft/TypeScript/issues/6716#issuecomment-226424254?

**Comment by zhengbli** [comment]
@chicoxyzzy Yes it does.

File: src/services/services.ts (hunk 1 above)
Comment (SaschaNaz) [inline src/services/services.ts]: Sorry but I'm not sure what you mean by 'optional', do you want me to make it `insert?: boolean` form?

**Comment by zhengbli** [comment]
@SaschaNaz Can you update the PR today? As we are getting closer to snap 2.0, this needs to get in soon

**Comment by SaschaNaz** [comment]
@zhengbli I can if I get the answer for https://github.com/Microsoft/TypeScript/pull/9361#discussion_r68694526. Not sure what 'optional' means here.

**Comment by mhegazy** [comment]
> Not sure what 'optional' means here as any other fields on FormatCodeOptions are not optional.

they were not optional originally, and that was a mistake. they should all have been optional from the get go.
Now adding a new one will break the build for any tool that uses formatting, even if they do not care to support this new option. so newer ones should be optional, and possibly we should go back and make all the options optional (no pun intended).

**Comment by SaschaNaz** [comment]
@mhegazy Got it. Should it also be optional on fourslash.ts and editorServices.ts or only on services.ts?

**Comment by SaschaNaz** [comment]
I think making it optional will make me forget adding corresponding fields on fourslash and editorServices for new default option value. Well, that is another story anyway ;)

**Comment by mhegazy** [comment]
thanks as always @SaschaNaz!

**Pull request merged**